src/lib/bit/bit.h

```diff
@@ -648,27 +663,157 @@ bit_clz_u64(uint64_t x)
 {
-#if   defined(HAVE_BUILTIN_POPCOUNT)
-	return __builtin_popcount(x);
-#else /* !defined(HAVE_BUILTIN_POPCOUNT) */
-	POPCOUNT_NAIVE(x, sizeof(uint32_t) * CHAR_BIT);
-#endif
+	return POPCOUNT(x);
 }
 
 /**
  * @copydoc bit_count_u32
  */
-inline int
+static inline int
 bit_count_u64(uint64_t x)
 {
-#if   defined(HAVE_BUILTIN_POPCOUNTLL)
-	return __builtin_popcountll(x);
-#else /* !defined(HAVE_BUILTIN_POPCOUNTLL) */
-	POPCOUNT_NAIVE(x, sizeof(uint64_t) * CHAR_BIT);
-#endif
+	return POPCOUNT(x);
+}
+
+/**
+ * @cond false
+ * @brief Return type of bitmap_word_align function.
+ */
+struct bitmap_word_align_params {
+	/**
+	 * Unaligned (we require word alignment)
+	 * bit prefix of the bitmap (cannot be processed
+	 * via popcount).
+	 */
+	int64_t leading_bits;
+	/**
+	 * Bit suffix of the bitmap which is smaller
+	 * than a word in size, which means popcount
+	 * cannot be used on it.
+	 */
+	int64_t trailing_bits;
+	/**
+	 * Bit offset from the start of the bitmap to
+	 * the start of the @a trailing_bits.
+	 */
+	int64_t trailing_bit_offset;
+	/**
+	 * Pointer to the start of the word aligned
+	 * part of the bitmap (can be processed via popcount).
+	 */
+	const uint8_t *aligned_start;
+	/** Count of bits in the word aligned part of the bitmap. */
+	int64_t aligned_bits;
+	/** Count of words in the word aligned part of the bitmap. */
+	int64_t aligned_words;
+};
+
+/**
+ * @brief Implementation of bitmap_word_align function.
+ */
+static inline struct bitmap_word_align_params
+bitmap_word_align(int64_t align_in_bytes, const uint8_t *data,
+		  int64_t bit_offset, int64_t length)
+{
+	assert(is_power_of_2(align_in_bytes));
+	const int64_t align_in_bits = align_in_bytes * 8;
+	/**
+	 * Compute a "bit address" that we can align up to
+	 * `align_in_bits`. We don't care about losing the
+	 * upper bits since we are only	interested in the
+	 * difference between both addresses.
+	 */
+	const uint64_t bit_addr = (size_t)data * 8 + (uint64_t)bit_offset;
+	const uint64_t aligned_bit_addr =
+		round_up_to_power_of_2(bit_addr, align_in_bits);
+	struct bitmap_word_align_params p;
+	p.leading_bits = MIN((uint64_t)length, aligned_bit_addr - bit_addr);
+	p.aligned_words = (length - p.leading_bits) / align_in_bits;
+	p.aligned_bits = p.aligned_words * align_in_bits;
+	p.trailing_bits = length - p.leading_bits - p.aligned_bits;
+	p.trailing_bit_offset = bit_offset + p.leading_bits + p.aligned_bits;
+	p.aligned_start = data + (bit_offset + p.leading_bits) / 8;
+	return p;
+}
+
+/** @endcond */
+
+/**
+ * @brief Returns the number of 1-bits in @a data vector starting from
+ *	  @a bit_offset.
+ * @param data byte vector of length @a length bits.
+ * @param bit_offset offset within the byte vector from where to start
+ *		     counting 1-bits.
+ * @param length length in bits of byte vector @a data.
+ * @return the number of 1-bits in @a data.
+ */
+static inline int64_t
+bit_count_vec(const uint8_t *data, int64_t bit_offset, int64_t length)
```

Comment:
We don't add `_vec` suffix to functions that deal with bitmaps of arbitrary length here (e.g. `bit_set`, `bit_test`). This one should probably be named simply `bit_count` for consistency.